src/metabase/driver/sql_jdbc/sync/describe_table.clj

```diff
@@ -349,56 +349,82 @@
-  100)
+  [value]
+  (if (and (string? value)
+           (can-parse-datetime? value))
+    java.time.LocalDateTime
+    (type value)))
+
+(defn- json-parser ^JsonParser [v]
+  (let [f (JsonFactory.)]
+    (if (string? v)
+      (.createParser f ^String v)
+      (.createParser f ^java.io.Reader v))))
+
+(defn- number-type [t]
+  (u/case-enum t
+    JsonParser$NumberType/INT Integer
+    JsonParser$NumberType/LONG Long
+    JsonParser$NumberType/FLOAT Float
+    JsonParser$NumberType/DOUBLE Double
+    JsonParser$NumberType/BIG_DECIMAL BigDecimal
+    JsonParser$NumberType/BIG_INTEGER clojure.lang.BigInt))
+
+(defn- json->types
+  "Parses given json (a string or a reader) into a map of paths to types, i.e. `{[\"bob\"} String}`.
+
+  Uses Jackson Streaming API to skip allocating data structures, eschews allocating values when possible.
+  Respects *nested-field-column-max-row-length*."
+  [v path]
+  (let [p (json-parser v)]
+    (loop [path (or path [])
+           key  nil
+           res  (transient {})]
+      (let [token (.nextToken p)]
+        (cond
+          (nil? token)
+          (persistent! res)
+
+          ;; we could be more precise here and issue warning about nested fields (the one in `describe-json-fields`),
+          ;; but this limit could be hit by multiple json fields rather than only by this one. So for the sake of
+          ;; issuing only a single warning in logs we'll spill over limit by a single entry (instead of doing `<=`).
+          (< max-nested-field-columns (count res))
+          (persistent! res)
+
+          :else
+          (u/case-enum token
+            JsonToken/VALUE_NUMBER_INT   (recur path key (assoc! res (conj path key) (number-type (.getNumberType p))))
+            JsonToken/VALUE_NUMBER_FLOAT (recur path key (assoc! res (conj path key) (number-type (.getNumberType p))))
+            JsonToken/VALUE_TRUE         (recur path key (assoc! res (conj path key) Boolean))
+            JsonToken/VALUE_FALSE        (recur path key (assoc! res (conj path key) Boolean))
+            JsonToken/VALUE_NULL         (recur path key (assoc! res (conj path key) nil))
+            JsonToken/VALUE_STRING       (recur path key (assoc! res (conj path key)
+                                                                 (type-by-parsing-string (.getText p))))
+            JsonToken/FIELD_NAME         (recur path (.getText p) res)
+            JsonToken/START_OBJECT       (recur (cond-> path key (conj key)) key res)
+            JsonToken/END_OBJECT         (recur (cond-> path (seq path) pop) key res)
+            ;; We put top-level array row type semantics on JSON roadmap but skip for now
+            JsonToken/START_ARRAY        (do (.skipChildren p)
+                                             (if key
+                                               (recur path key (assoc! res (conj path key) clojure.lang.PersistentVector))
+                                               (recur path key res)))
```

Comment:
The map returned in end is going to be pretty ambiguous (empty map or any array). Semantics I guess are fine for now though.

Reply:
If I understand correctly, all of the following with return an empty map:

```json
{}
[]
[{"status": "forgotten"}]
[1, 2, 3]
```

Reply:
Not saying this is a problem, just a bit opaque for debugging.

Reply:
Eh, pepper code with `#p` and you'll know what's going on in no time. :)

Reply:
Hehe I mean debugging in production, but thanks for the reminder I forgot about that awesome reader macro.